sql/driver/driver_test.go

```diff
@@ -258,3 +259,100 @@ func TestInsecure(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// concurrentIncrements starts two Goroutines in parallel, both of which
+// read the integer stored at the other's key, increment and update their own.
+// It checks that the outcome is serializable, i.e. exactly one of the
+// two Goroutines (the later write) sees the previous write by the other.
+func concurrentIncrements(db *sql.DB, t *testing.T) {
+	var wgStart, wgEnd sync.WaitGroup
+	wgStart.Add(2)
+	wgEnd.Add(2)
+	for i := 0; i < 2; i++ {
+		go func(i int) {
+			defer wgEnd.Done()
+			shouldStartWait := true
+			// Loop until success.
+			for {
+				txn, err := db.Begin()
+				if err != nil {
+					t.Fatal(err)
+				}
+				// Although the SELECT and the UPDATE below can be combined into a
+				// single statement, we prefer them being separate to ensure that
+				// this transaction is truly a multi-statement transaction, providing
+				// plenty of opportunity to mess up serializability.
+				var value int64
+				err = txn.QueryRow(`SELECT v FROM t.kv WHERE k = $1`, (i+1)%2).Scan(&value)
+				if err != nil {
+					_ = txn.Rollback()
+					continue
+				}
+				// Synchronization point to ensure that there is contention.
+				if shouldStartWait {
```

Comment:
I did add in log messages to check how the code runs and its pretty unpredictable. Usually the second read immediately follows the first read but there is no guarantee that's always the case.
Also using wgStart doesn't seem to have any effect. So I'd definitely recommend we take that out.

I think the only remaining issue here is whether to leave it as is or to deliberately ensure that the second read follows the first which is how I had it earlier. I'm in favor of adding in the wait after the read.